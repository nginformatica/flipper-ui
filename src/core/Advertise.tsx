import React, { FC } from 'react'
import { background } from '../colors'
import MuiPaper from './Paper'
import Typography from './Typography'
import styled from 'styled-components'
import type { DefaultProps } from './types'

interface AdvertiseProps extends DefaultProps {
    comment: string
    author: string
    commentStyle?: object
    authorStyle?: object
}

const styles = {
    border: {
        borderLeft: `1px solid ${background.dark}`
    }
}

const Paper = styled(MuiPaper)`
    display: flex;
    flex-direction: column;
`

const Advertise: FC<AdvertiseProps> = ({
    comment,
    author,
    padding = 4,
    commentStyle = {},
    authorStyle = {},
    ...otherProps
}) => (
    <Paper padding={padding} {...otherProps}>
        <Typography
            margin='0 12px'
            padding='6px 18px'
            style={{ ...styles.border, ...commentStyle }}>
            {comment}
        </Typography>
        <Typography
            margin='0px 12px'
            padding='6px 18px'
            variant='caption'
            style={{ ...styles.border, ...authorStyle }}>
            {author}
        </Typography>
    </Paper>
)

export default Advertise
