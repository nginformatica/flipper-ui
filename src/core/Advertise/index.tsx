import React from 'react'
import { background } from '../../colors'
import MuiPaper from '../Paper'
import Typography from '../Typography'
import styled from 'styled-components'
import type { DefaultProps } from '../types'

export interface AdvertiseProps extends DefaultProps {
    comment: string
    author: string
    commentStyle?: object
    authorStyle?: object
}

const styles = {
    border: {
        borderLeft: `1px solid ${background.dark}`
    }
}

const Paper = styled(MuiPaper)`
    display: flex;
    flex-direction: column;
`

const Advertise = ({
    comment,
    author,
    padding = 4,
    commentStyle = {},
    authorStyle = {},
    ...otherProps
}: AdvertiseProps) => (
    <Paper padding={padding} {...otherProps}>
        <Typography
            margin='0 12px'
            padding='6px 18px'
            style={{ ...styles.border, ...commentStyle }}>
            {comment}
        </Typography>
        <Typography
            margin='0px 12px'
            padding='6px 18px'
            variant='caption'
            style={{ ...styles.border, ...authorStyle }}>
            {author}
        </Typography>
    </Paper>
)

export default Advertise
