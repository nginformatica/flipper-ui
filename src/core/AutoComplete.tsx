import Downshift from 'downshift'
import {
    contains,
    filter,
    is,
    isNil,
    pipe,
    propEq,
    propOr,
    toLower,
    uniq,
    unless,
    when
} from 'ramda'
import React, {
    cloneElement,
    Component,
    FocusEvent,
    Fragment,
    ReactElement,
    ReactNode
} from 'react'
import Paper from './Paper'

interface IProps {
    data: object[] | string[]
    value: string
    defaultValue?: string
    inputElement: ReactElement<any>
    InputProps?: object
    defaultIsOpen?: boolean
    openOnFocus?: boolean
    renderSuggestion: (
        suggestion: string | object,
        itemProps: object,
        selected: boolean
    ) => ReactNode
    onChange?: (value: string) => void
    onSelect?: (value: string) => void
    onFocus?: (event: FocusEvent<HTMLInputElement>) => void
    onBlur?: (event: FocusEvent<HTMLInputElement>) => void
}

class AutoComplete extends Component<IProps> {
    public autocomplete = { offsetWidth: 256 }
    public getSuggestions(inputValue) {
        if (this.props.openOnFocus && !inputValue) {
            return this.props.data
        }

        const items = inputValue
            ? this.props.data
            : []

        return uniq(
            filter(
                unless(
                    propEq('subheader', true),
                    pipe(
                        when(is(Object), propOr('', 'label')),
                        toLower,
                        contains(toLower(inputValue || ''))
                    )
                ),
                items
            )
        )
    }

    public handleSelect(value) {
        if (this.props.onSelect) {
            this.props.onSelect(value)
        } else {
            this.handleChange(value)
        }
    }

    public handleChange(value) {
        if (!isNil(value) && this.props.onChange) {
            this.props.onChange(value)
        }
    }

    public renderInput(props: {
        inputProps: object,
        onFocus: (event: FocusEvent<HTMLInputElement>) => void
    }) {
        return cloneElement(this.props.inputElement, {
            InputProps: {
                ...props,
                ...this.props.InputProps
            },
            inputProps: {
                ref: self => {
                    this.autocomplete = self
                }
            }
        })
    }

    public render() {
        const paperStyle = {
            position: 'absolute',
            width: this.autocomplete.offsetWidth,
            zIndex: 1099
        }

        return (
            <Downshift
                inputValue={ this.props.value }
                defaultInputValue={ this.props.defaultValue }
                defaultIsOpen={ this.props.defaultIsOpen }
                itemToString={ item => is(Object, item) ? item.label : item }
                onSelect={ this.handleSelect.bind(this) }
                onInputValueChange={ this.props.onChange }>
                {
                    ({ isOpen, getInputProps, inputValue, getItemProps, highlightedIndex, openMenu }) =>
                        <div>
                            {
                                this.renderInput(getInputProps({
                                    onBlur: this.props.onBlur,
                                    onFocus: (event: FocusEvent<HTMLInputElement>) => this.props.onFocus
                                        ? this.props.onFocus(event)
                                        : this.props.openOnFocus
                                            ? openMenu()
                                            : null,
                                    value: inputValue || ''
                                }))
                            }
                            {
                                isOpen && (
                                    <Paper square style={ paperStyle }>
                                        {
                                            this.getSuggestions(inputValue).map((suggestion, index) =>
                                                <Fragment key={ index }>
                                                    {
                                                        this.props.renderSuggestion(
                                                            suggestion,
                                                            getItemProps({ item: suggestion }),
                                                            highlightedIndex === index
                                                        )
                                                    }
                                                </Fragment>
                                            )
                                        }
                                    </Paper>
                                )
                            }
                        </div>
                }
            </Downshift>
        )
    }
}

export default AutoComplete
