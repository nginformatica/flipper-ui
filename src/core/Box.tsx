import React, { FC } from 'react'
import type { DefaultProps } from './types'
import Paper from './Paper'

interface BoxProps extends DefaultProps {
    minHeight?: number
}

const Box: FC<BoxProps> = ({
    children,
    margin,
    padding = 18,
    style = {},
    minHeight = 400,
    ...otherProps
}) =>
    <Paper
        padding={ padding }
        margin={ margin }
        style={ { minHeight, ...style } }
        { ...otherProps }>
        { children }
    </Paper>

export default Box
