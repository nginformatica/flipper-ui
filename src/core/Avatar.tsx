import { Avatar as MuiAvatar } from '@material-ui/core'
import { makeStyles, createStyles } from '@material-ui/core/styles'
import { Theme } from '@material-ui/core/styles/createMuiTheme'
import React, { FC } from 'react'
import type { DefaultProps } from './types'

interface AvatarProps extends DefaultProps {
    name?: string
    id?: string
    sizes?: string
    src?: string
    imgProps?: object
    primary?: boolean
}

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        primary: {
            backgroundColor: theme.palette.primary.main
        }
    })
)

const Avatar: FC<AvatarProps> = ({
    children,
    primary,
    className,
    ...otherProps
}) => {
    const classes = useStyles()

    return (
        <MuiAvatar
            { ...otherProps }
            className={ `${className} ${primary ? classes['primary'] : ''}` }>
            { children }
        </MuiAvatar>
    )
}

export default Avatar
