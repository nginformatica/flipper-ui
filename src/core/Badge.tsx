import { Badge as MuiBadge } from '@material-ui/core'
import React, { FC } from 'react'
import { BadgeProps as MuiBadgeProps } from '@material-ui/core/Badge'
import { makeStyles } from '@material-ui/core/styles'
import type { DefaultProps } from './types'

export interface BadgeProps extends DefaultProps, MuiBadgeProps {
    max?: number
    counter: number | string
    position?: {
        top?: number
        bottom?: number
        left?: number
        right?: number
    }
}

const useBadgeStyles = (position: BadgeProps['position']) => {
    const getStyles = makeStyles({
        badge: {
            ...position
        }
    })

    return getStyles()
}

const Badge: FC<BadgeProps> = props => {
    const {
        children,
        counter,
        padding,
        margin,
        position,
        style = {},
        ...otherProps
    } = props

    const classes = useBadgeStyles(position)

    return counter
        ? (
            <MuiBadge
                badgeContent={ counter }
                classes={ classes }
                style={ { padding, margin, ...style } }
                { ...otherProps }>
                { children }
            </MuiBadge>
        )
        : <>{ children }</>
}

export default Badge
