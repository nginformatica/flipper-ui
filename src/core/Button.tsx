import { Button as MuiButton } from '@material-ui/core'
import React, { ElementType, MouseEvent } from 'react'
import styled from 'styled-components'
import { DefaultProps } from './types'

export interface ButtonProps extends DefaultProps {
    disabled?: boolean
    selected?: boolean
    disableTouchRipple?: boolean
    component?: ElementType
    color?: 'default' | 'primary' | 'inherit' | 'secondary'
    size?: 'small' | 'medium' | 'large'
    href?: string
    fullWidth?: boolean
    variant?: 'text' | 'outlined' | 'contained' | 'dashed'
    target?: string
    children?: React.ReactNode
    'data-testid'?: string
    onClick?(event: MouseEvent<HTMLButtonElement>): void
}

const StyledButton = styled(MuiButton)<
    ButtonProps & { dashed?: 'true' | 'false' }
>`
    border-style: ${props =>
        props.dashed === 'true' ? 'dashed !important' : 'initial'};
    opacity: ${props => (props.selected ? 0.5 : 1)};
`

const Button = ({
    children,
    margin,
    padding,
    style = {},
    variant,
    ...otherProps
}: ButtonProps) => (
    <StyledButton
        {...otherProps}
        dashed={variant === 'dashed' ? 'true' : 'false'}
        variant={variant === 'dashed' ? 'outlined' : variant}
        style={{ margin, padding, ...style }}>
        {children}
    </StyledButton>
)

export default Button
