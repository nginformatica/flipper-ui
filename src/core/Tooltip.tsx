import { Tooltip as MuiTooltip } from '@material-ui/core'
import React, { ReactElement, CSSProperties } from 'react'
import { IDefault } from './Advertise'

interface IProps extends IDefault {
    placement?:
        | 'bottom-end'
        | 'bottom-start'
        | 'bottom'
        | 'left-end'
        | 'left-start'
        | 'left'
        | 'right-end'
        | 'right-start'
        | 'right'
        | 'top-end'
        | 'top-start'
        | 'top'
    title: React.ReactNode
    withWrapper?: boolean
    wrapperStyle?: CSSProperties
    onClose?: () => void
    onOpen?: () => void
    open?: boolean
    children: ReactElement<{}>
    enterDelay?: number
}

const Tooltip = ({
    children,
    withWrapper,
    wrapperStyle,
    enterDelay = 1000,
    title,
    ...otherProps
}: IProps) =>
    <MuiTooltip title={ title || '' } enterDelay={ enterDelay } { ...otherProps }>
        {
            withWrapper
                ? (
                    <div
                        style={ {
                            display: 'inherit',
                            ...wrapperStyle
                        } }>
                        { children }
                    </div>
                )
                : children
        }
    </MuiTooltip>

export default Tooltip
