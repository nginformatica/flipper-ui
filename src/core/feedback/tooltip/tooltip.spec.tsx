import React from 'react'
import { act } from 'react-dom/test-utils'
import { Button } from '@material-ui/core'
import { render, screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import Tooltip from '.'

describe('Tooltip', () => {
    it('should render a Tooltip', async () => {
        render(
            <Tooltip enterDelay={1} title='Tooltip Message'>
                <Button variant='outlined'>Tooltip</Button>
            </Tooltip>
        )

        const label = screen.getByText('Tooltip')

        await act(async () => {
            await userEvent.hover(label)
        })

        const message = screen.getByText('Tooltip Message')

        expect(message).toBeDefined()
    })

    it('should render a Tooltip with wrapper', async () => {
        render(
            <Tooltip withWrapper enterDelay={1} title='Tooltip Message'>
                <Button variant='outlined'>Tooltip</Button>
            </Tooltip>
        )

        const label = screen.getByText('Tooltip')

        await act(async () => {
            await userEvent.hover(label)
        })

        const message = screen.getByText('Tooltip Message')

        expect(message).toBeDefined()
    })

    it('should not render a Tooltip', async () => {
        render(
            <Tooltip title=''>
                <Button variant='outlined'>Tooltip</Button>
            </Tooltip>
        )

        const label = screen.getByText('Tooltip')

        await act(async () => {
            await userEvent.hover(label)
        })

        const message = screen.queryByText('Tooltip Message')

        expect(message).toBeNull()
    })

    it('should match snapshot', () => {
        const { container } = render(
            <Tooltip withWrapper enterDelay={1} title='Tooltip Message'>
                <Button variant='outlined'>Tooltip</Button>
            </Tooltip>
        )

        expect(container).toMatchSnapshot()
    })
})
