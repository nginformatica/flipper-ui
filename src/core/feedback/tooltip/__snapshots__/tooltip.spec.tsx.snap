// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should match snapshot 1`] = `
<div>
  <div
    aria-label="Tooltip Message"
    class=""
    data-mui-internal-clone-element="true"
    style="display: inherit;"
  >
    <button>
      Tooltip
    </button>
  </div>
</div>
`;
